mapeditor/inspector/townspellswidget.cpp

```diff
@@ -0,0 +1,172 @@
+/*
+ * townspellswidget.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#include "StdInc.h"
+#include "townspellswidget.h"
+#include "ui_townspellswidget.h"
+#include "inspector.h"
+#include "../../lib/constants/StringConstants.h"
+#include "../../lib/spells/CSpellHandler.h"
+
+TownSpellsWidget::TownSpellsWidget(CGTownInstance & town, QWidget * parent) :
+	QDialog(parent),
+	ui(new Ui::TownSpellsWidget),
+	town(town)
+{
+	ui->setupUi(this);
+	BuildingID mageGuilds[] = {BuildingID::MAGES_GUILD_1, BuildingID::MAGES_GUILD_2, BuildingID::MAGES_GUILD_3, BuildingID::MAGES_GUILD_4, BuildingID::MAGES_GUILD_5};
+	for (int i = 0; i < GameConstants::SPELL_LEVELS; i++)
+	{
+		ui->tabWidget->setTabEnabled(i, vstd::contains(town.getTown()->buildings, mageGuilds[i]));
+	}
+}
+
+TownSpellsWidget::~TownSpellsWidget()
+{
+	delete ui;
+}
+
+
+void TownSpellsWidget::obtainData()
+{
+	initSpellLists();
+	if (vstd::contains(town.possibleSpells, SpellID::PRESET)) {
+		ui->customizeSpells->setChecked(true);
+	}
+	else
+	{
+		ui->customizeSpells->setChecked(false);
+		ui->tabWidget->setEnabled(false);
+	}
+}
+
+void TownSpellsWidget::resetSpells()
+{
+	town.possibleSpells.clear();
+	town.obligatorySpells.clear();
+	for (auto spell : VLC->spellh->objects)
+	{
+		if (!spell->isSpecial() && !spell->isCreatureAbility())
+			town.possibleSpells.push_back(spell->id);
+	}
+}
+
+void TownSpellsWidget::initSpellLists()
+{
+	QListWidget * possibleSpellLists[] = { ui->possibleSpellList1, ui->possibleSpellList2, ui->possibleSpellList3, ui->possibleSpellList4, ui->possibleSpellList5 };
+	QListWidget * requiredSpellLists[] = { ui->requiredSpellList1, ui->requiredSpellList2, ui->requiredSpellList3, ui->requiredSpellList4, ui->requiredSpellList5 };
+	auto spells = VLC->spellh->objects;
```

Comment:
`VLC->spellh->getDefaultAllowed()` is preferred to direct access to `objects` field.

And to convert SpellID to Spell you can use (preferred) `spellID.toEntity(VLC)` or `spellID.toSpell()` (if you need info not present in entity)